src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -432,17 +440,24 @@ private void AddDirectoryWatchUnlocked(WatchedDirectory? parent, string director
                 // asked for subdirectories to be included.
                 if (isNewDirectory && _includeSubdirectories)
                 {
-                    // This method is recursive.  If we expect to see hierarchies
-                    // so deep that it would cause us to overflow the stack, we could
-                    // consider using an explicit stack object rather than recursion.
-                    // This is unlikely, however, given typical directory names
-                    // and max path limits.
-                    foreach (string subDir in Directory.EnumerateDirectories(fullPath))
+                    try
                     {
-                        AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(subDir));
-                        // AddDirectoryWatchUnlocked will add the new directory to
-                        // this.Children, so we don't have to / shouldn't also do it here.
+                        // This method is recursive.  If we expect to see hierarchies
+                        // so deep that it would cause us to overflow the stack, we could
+                        // consider using an explicit stack object rather than recursion.
+                        // This is unlikely, however, given typical directory names
+                        // and max path limits.
+                        foreach (string subDir in Directory.EnumerateDirectories(fullPath))
+                        {
+                            AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(subDir));
+                            // AddDirectoryWatchUnlocked will add the new directory to
+                            // this.Children, so we don't have to / shouldn't also do it here.
+                        }
                     }
+                    catch (DirectoryNotFoundException)
+                    { } // The child directory was removed.
+                    catch (IOException ex) when (ex.HResult == Interop.Error.ENOTDIR.Info().RawErrno)
+                    { } // The child directory was replaced by a file.
```

Comment:
if we stop on the first error we might omit some valid directories that were not enumerated yet. This would be a silent error.

The alternative is to materialize all the directories at once and then iterate over them. Sth like this:

```cs
List<string> directories = new List<string>(Directory.GetDirectories("path"));

for (int i = 0; i < directories.Count; i++)
{
    AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(directories[i]));
}
```

The disadvantage is that we would allocate more memory. I am also not sure whether `Directory.GetDirectories` can fail in such scenario or not.

@JeremyKuhne @carlossanlop do you know if there is any other API for enumerating directories that does not throw when one of the directories gets removed?

Reply:
@adamsitnik you might be able to write a custom `FileSystemEnumerator` and ignore the not found errors. This sort of thing might be a useful new enumeration option as well...

Reply:
> do you know if there is any other API for enumerating directories that does not throw when one of the directories gets removed?

`EnumerateDirectories` throws for the path you pass in. It doesn't throw during enumeration.